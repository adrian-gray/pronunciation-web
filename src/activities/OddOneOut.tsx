import React, { useState } from "react";

import Table from "react-bootstrap/Table";

import Cell from "./../components/Cell";
import MemberGate from "./../components/MemberGate";
import SplitHilite from "./../components/SplitHilite";

interface IOddOneOutProps {
  isUserAuth: boolean;
  tag: string;
  ipa: string;
  example: string[];
  exampleHilite: number;
  rows: string[][];
  correct: string[];
}

export default (props: IOddOneOutProps) => {
  const title = `Select the words that ~DON'T~ have the ${props.tag} – ${props.ipa} sound.`;

  const [rowHiliteColor, setRowHiliteColor] = useState(
    new Array(props.rows.length)
  );
  const [selected, setSelected] = useState(
    new Array(props.rows.length).fill(-1)
  );

  function buildTable() {
    const displayRows = props.rows.map((arr, row) => {
      const cells = arr.map((word, column) => {
        const isSelected = selected[row] === column;
        const cellBG = isSelected ? rowHiliteColor[row] : null;
        return (
          <Cell
            str={word}
            column={column}
            row={row}
            handleClick={handleClick}
            hilite={cellBG}
            key={word}
          />
        );
      });
      return <tr key={row}>{cells}</tr>;
    });

    const guide = props.example.map((word, i) => {
      const hilite = i + 1 === props.exampleHilite ? "selected-bg" : null;
      // TODO why am I passing extra data here?
      return (
        <Cell
          str={word}
          key={i}
          hilite={hilite}
          row={null}
          column={null}
          handleClick={(): void => null}
        />
      );
    });

    return (
      <div>
        <h4 className="head=space text-center">{"Example"}</h4>
        <Table>
          <thead className="th-bg">
            <tr>{guide}</tr>
          </thead>
          <tbody>{displayRows}</tbody>
        </Table>
      </div>
    );
  }

  function handleClick(params: { column: number; row: number }) {
    if (!props.isUserAuth) return;
    const { column, row } = params;
    const arr = selected.slice();
    arr[row] = column;
    setSelected(arr);

    checkComplete(arr);
  }

  function checkComplete(arr: number[]) {
    const hiliteColor = rowHiliteColor.slice();
    arr.forEach((column: number, row: number) => {
      if (column > -1) {
        const word = props.rows[row][column];
        if (props.correct.includes(word)) {
          hiliteColor[row] = true;
        } else {
          hiliteColor[row] = false;
        }
      }
    });
    setRowHiliteColor(hiliteColor);
  }

  const content = buildTable();

  return (
    <div className="head-space">
      <h3>
        {`The Odd One Out using ${props.tag} - `}
        <SplitHilite str={props.ipa} />
      </h3>
      <h4>
        <SplitHilite str={title} />
      </h4>
      <br />
      <p>
        {`Can you spot the difference? In each group of words there are three with the ${props.tag} vowel sound and one with a different vowel sound. Say the words and choose the one that doesn’t sound the same as the others. Advance to the next level with more challenging words. Odd One Out encourages you to identify the differences between words with similar sounding vowels.`}
      </p>
      <MemberGate content={content} isUserAuth={props.isUserAuth} />
    </div>
  );
};
