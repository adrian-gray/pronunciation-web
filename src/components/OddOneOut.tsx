import React, { useState } from "react";

import Table from "react-bootstrap/Table";

import Cell from "./Cell";
import MemberGate from "./MemberGate";
import SplitHilite from "./SplitHilite";

export default props => {
  const title = `Select the words that ~DON'T~ have the ${props.tag} – ${props.ipa} sound.`;

  const [rowHiliteColor, setRowHiliteColor] = useState(
    new Array(props.rows.length)
  );
  const [selected, setSelected] = useState(
    new Array(props.rows.length).fill(-1)
  );

  function buildTable() {
    const displayRows = props.rows.map((arr, row) => {
      const cells = arr.map((word, column) => {
        const isSelected = selected[row] === column;
        const cellBG = isSelected ? rowHiliteColor[row] : null;
        return (
          <Cell
            str={word}
            column={column}
            row={row}
            handleClick={handleClick}
            hilite={cellBG}
          />
        );
      });
      return <tr key={row}>{cells}</tr>;
    });

    const guide = props.example.map((word, i) => {
      const hilite = i + 1 === props.exampleHilite ? "selected-bg" : null;
      return <Cell str={word} key={i} hilite={hilite} />;
    });

    return (
      <div>
        <h4 className="head=space text-center">{"Example"}</h4>
        <Table>
          <thead className="th-bg">
            <tr>{guide}</tr>
          </thead>
          <tbody>{displayRows}</tbody>
        </Table>
      </div>
    );
  }

  function handleClick(params) {
    if (!props.userAuth) return;
    const { column, row } = params;
    const arr = selected.slice();
    arr[row] = column;
    setSelected(arr);

    checkComplete(arr);
  }

  function checkComplete(arr) {
    const hiliteColor = rowHiliteColor.slice();
    arr.forEach((column, row) => {
      if (column > -1) {
        const word = props.rows[row][column];
        if (props.correct.includes(word)) {
          hiliteColor[row] = "correct-bg";
        } else {
          hiliteColor[row] = "incorrect-bg";
        }
      }
    });
    setRowHiliteColor(hiliteColor);
  }

  const content = buildTable();

  return (
    <div className="head-space">
      <h3>
        {`The Odd One Out using ${props.tag} - `}
        <SplitHilite str={props.ipa} />
      </h3>
      <h4>
        <SplitHilite str={title} />
      </h4>
      <br />
      <p>
        {`Can you spot the difference? In each group of words there are three with the ${props.tag} vowel sound and one with a different vowel sound. Say the words and choose the one that doesn’t sound the same as the others. Advance to the next level with more challenging words. Odd One Out encourages you to identify the differences between words with similar sounding vowels.`}
      </p>
      <MemberGate content={content} userAuth={props.userAuth} />
    </div>
  );
};
