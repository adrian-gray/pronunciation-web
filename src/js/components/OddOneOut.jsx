import React from 'react'

export default function OddOneOut (props) {
  const words = props.words.map((word, index) => {
    return <div key={index} className='col'>{word}</div>
  })

  return (
    <div className='container'>
      <div className='row'>
        {words}
      </div>
    </div>
  )
}
