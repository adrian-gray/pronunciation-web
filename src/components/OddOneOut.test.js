import React from "react";
import { render, cleanup } from "@testing-library/react";

import OddOneOut from "./OddOneOut";

afterEach(cleanup);

test("OddOneOut", async () => {
  const props = {
    ipa: "a:",
    tag: "long-a",
    rows: [["the", "cat", "sat"], ["on", "the", "mat"]],
    correct: ["the", "on"],
    example: ["three", "blind", "moice"],
    exampleHilite: 3,
    isUserAuth: true
  };

  const { container } = render(<OddOneOut {...props} />);

  expect(container).toMatchSnapshot();
});
