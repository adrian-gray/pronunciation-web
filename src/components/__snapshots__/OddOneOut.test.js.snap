// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OddOneOut 1`] = `
<div>
  <div
    class="head-space"
  >
    <h3>
      The Odd One Out using long-a - 
      <span>
        a:
      </span>
       
    </h3>
    <h4>
      <span>
        Select the words that 
      </span>
      <span
        class="hilite"
      >
        DON'T
      </span>
      <span>
         have the long-a – a: sound.
      </span>
       
    </h4>
    <br />
    <p>
      Can you spot the difference? In each group of words there are three with the long-a vowel sound and one with a different vowel sound. Say the words and choose the one that doesn’t sound the same as the others. Advance to the next level with more challenging words. Odd One Out encourages you to identify the differences between words with similar sounding vowels.
    </p>
    <div>
      <h4
        class="head=space text-center"
      >
        Example
      </h4>
      <table
        class="table"
      >
        <thead
          class="th-bg"
        >
          <tr>
            <td
              class="cell"
            >
              <p>
                <span>
                  three
                </span>
                 
              </p>
            </td>
            <td
              class="cell"
            >
              <p>
                <span>
                  blind
                </span>
                 
              </p>
            </td>
            <td
              class="cell hilite"
            >
              <p>
                <span>
                  moice
                </span>
                 
              </p>
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="cell"
            >
              <p>
                <span>
                  the
                </span>
                 
              </p>
            </td>
            <td
              class="cell"
            >
              <p>
                <span>
                  cat
                </span>
                 
              </p>
            </td>
            <td
              class="cell"
            >
              <p>
                <span>
                  sat
                </span>
                 
              </p>
            </td>
          </tr>
          <tr>
            <td
              class="cell"
            >
              <p>
                <span>
                  on
                </span>
                 
              </p>
            </td>
            <td
              class="cell"
            >
              <p>
                <span>
                  the
                </span>
                 
              </p>
            </td>
            <td
              class="cell"
            >
              <p>
                <span>
                  mat
                </span>
                 
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
